docs/browser/events.md

```diff
@@ -0,0 +1,158 @@
+# Semantic conventions for browser events
+
+**Status**: [Experimental](../../../document-status.md)
+
+<details>
+<summary>Table of Contents</summary>
+<!-- Re-generate TOC with `markdown-toc --no-first-h1 -i` -->
+
+<!-- toc -->
+
+- [PageView](#pageview)
+- [PageNavigationTiming](#pagenavigationtiming)
+- [ResourceTiming](#resourcetiming)
+- [Exception](#exception)
+- [UserAction](#useraction)
+- [WebVital](#webvital)
+
+<!-- tocstop -->
+
+</details>
+
+This document describes the semantic conventions for events on browser platform. All browser events MUST use a namespace of `browser` in the `event.name` property.
+
+## PageView
+
+The event name MUST be `browser.page_view`.
+
+This event describes the details of the web page visited.
+
+The following table describes the payload fields that MUST be used to describe the details of event.
+
+| Body Field  | Type | Description  | Examples  | Requirement Level |
+|---|---|---|---|---|
+| `referrer` | string | Referring Page URI (`document.referrer`) whenever available. | `https://en.wikipedia.org/wiki/Main_Page` | Recommended |
+| `type` | int | Browser page type | `0` | Required |
+| `title` | string | Page title DOM property | `Shopping cart page` | Recommended |
+| `url` [1] | string | Full HTTP request URL in the form `scheme://host[:port]/path?query[#fragment]`. Usually the fragment is not transmitted over HTTP, but if it is known, it should be included nevertheless. [2] | `https://en.wikipedia.org/wiki/Main_Page`; `https://en.wikipedia.org/wiki/Main_Page#foo` | Required |
+| `changeState` | string | Type of state change used for the virtual page navigation | `pushState`, `replaceState` | Recommended |
+
+**[1]:**  Alias for [`http.url`](../../../trace/semantic_conventions/http.md)
+**[2]:** The URL fragment may be included for virtual pages
+
+`type` MUST be one of the following:
+
+| Value  | Description |
+|---|---|
+| `0` | physical_page - Initial page load within the browser and will generally also precede a PageNavigationTiming event.|
```

Comment:
**Question:** Is there a way to identify everything belonging to the same `physical_page`? For example, to re-create a browser dev tools network tab based on this data?

Reply:
@bripkens the `name` field of `browser.page_navigation_timing` would have the URL of the page, and would match the `url` field of `browser.page_view`.

Reply:
@scheler that implies that name is used as an approximation for correlation purposes, i.e., two tabs on the same URL could not be differentiated?

Reply:
@bripkens Applications should also setup an SDK processor that adds [session.id](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/general/session.md) attribute to each outgoing event. With this, different tabs would get different `session.id` which can then be used for the differentiation you describe.

Let me sync up with @MSNev and see how I can add this attribute to the browser events definition.

Reply:
@scheler that implies that a user interacting with a website would have to result in multiple concurrently valid `session.id`. Which seems counter-intuitive. 

I would propose to introduce an attribute that ties all events for a single physical page load together. That way, it is possible to differentiate between web browser tabs and to reconstruct a dev tool-like experience based on OTel data.

Reply:
There's resource attribute [`service.instance.id`](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/attributes-registry/service.md#service-attributes) which can be interpreted to cover this (each page load being a separate instance with new id generated on SDK init). There's been some arguments about if service namespace is correct for apps but imo that's such pedantic non-issue - rename it if must but concept is same

Reply:
@bripkens I checked the Client SIG [notes](https://docs.google.com/document/d/16Vsdh-DM72AfMg_FIt9yT9ExEWF4A_vRbQ3jRNBe09w/edit) and on 11/14/2023 we had discussed an attribute `browser.page.instance_id` which can be used for the purpose you are suggesting. I can add that, however,  if we want to put it inside the event payloads it can be further shorter non-namespaced name, say `page_id`.

Reply:
Added `browser.page.instance_id` as attribute on the Event, as at this point it's not clear what the semantics are for a processor to add payload fields.

Reply:
> I would imagine this to be confusing for an OTel end-user. A web developer working with/seeing OTel data would imho not be able to interpret this.

I don't think the issue exists as much as you imply here. For one a dev would end up defining app level information such as app name, version, environment... And so far all conversations regarding describing web apps have ended up as apps are services (see https://github.com/open-telemetry/opentelemetry-specification/issues/3791), therefore the user would have to already been aware of resources and the `service.` namespace (`service.name`, `service.version`, `deployment.environment`)

And at this point I'd just might as well copy-paste @svrnm's comment from https://github.com/open-telemetry/opentelemetry-specification/issues/3791#issuecomment-1867600136 disagreeing with your opinion and highlighting consistency between backend and frontend modelling



So next up, `browser.page.instance_id` attribute on Event has the issue that it gets duplicated on every outgoing signal, which is a waste of resources even when compressed (and y'know RUM should be minimal impact to user's device and bandwidth). And as it doesn't change after being set once, it's ideal candidate to be on resource. You know what's also on resource? Yep. So I don't see any value in this addition

Reply:
I will not deny that existing attributes are well-defined and designed for usage on a resource level. And, of course, I am all in favor of reusing existing semantics. To be transparent, I haven't been involved in any other discussions around this, so my context here is limited.

The point I was trying to make is merely about end-users' ability to interpret the attributes they see. For service name, version, and environment, I (personally/subjective) find the similarities between backend/frontend much more straightforward than for the case we are discussing.

My statement is under the assumption that end-users would see those attributes and need to interpret them to some degree.

Having some means to correlate activity to a physical page load/hard navigation was my core objective, though. That said, if there is some documentation on the meaning of `service.intance.id` for RUM, then I would be happy to rest my case :).

Reply:
Events (may) be tagged with the current trace / span, so for applications that have a current span this will always be unique to the current page load -- so we already have that ability (if required).

Failing that as @scheler has called out each "page" view will for the same "user" in the same "session" will be tagged with the current session.id, and the URL can be used to tie it together.

So a combination of the URL, session.id, (anonymous user id (under discussion in semconv)) and the actual "time" of the event as the "time" is the current browser runtime time (so there is very little chance of clock drift -- beyond an NTP time reset occurring while the page is loading).

So i short, we dont need another "linking" id / attribute, we SHOULD use the current traceid (and potentially the spanid) as the traceid for every page load is unique (or should be)

Reply:
And that does NOT need to be defined as part of the specific event implementation, as it SHOULD be part of the general reporting of the events (ie. the Event SDK implementation -- which reuses the Log implementation).

Reply:
We discussed in the Event SIG today that the event schema will not include fields that are added later, say by processors, after the events are emitted by the instrumentation. Since the field `browser.page.instance_id` will be added by an SDK processor to all the events emitted on a page instance, I will remove it from the Event definition here.